### PR TITLE
dcc editor: animation player

### PR DESCRIPTION
**Pull request by gucio321**
Hi there,
I got an idea, to create animation player for our dcc/dc6 widgets.
In connection with `change palette` feature, it allows users to  view an animation.

Let me know, if `time.Ticker` isn't a good solution.

### CODE REVIEW

File: hswidget/custom_widgets.go [hunk 1]
```diff
@@ -25,6 +30,123 @@ func MakeImageButton(id string, w, h int, t *giu.Texture, fn func()) giu.Widget
...
+	pauseTexture *giu.Texture
+}
+
+func (s *playPauseButtonState) Dispose() {
+	s.playTexture = nil
+	s.playTexture = nil
+}
+
+// PlayPauseButtonWidget represents a play/pause button
+type PlayPauseButtonWidget struct {
```
Comment (gucio321) [inline hswidget/custom_widgets.go:44]: btw, we can use it in `hssoundeditor`

File: hswidget/dccwidget/state.go [hunk 2]
```diff
@@ -5,20 +5,61 @@ import (
...
+	playModeForward animationPlayMode = iota
+	playModeBackword
+	playModeLeftRight
 )
 
+func (a animationPlayMode) String() string {
+	s := map[animationPlayMode]string{
+		playModeForward:   "Play Forward",
+		playModeBackword:  "Play Backword",
+		playModeLeftRight: "Play forward and than backword",
```
Comment (gucio321) [inline hswidget/dccwidget/state.go]: image

File: hswidget/dc6widget/state.go [hunk 3]
```diff
@@ -5,12 +5,41 @@ import (
...
+	miliseconds     = 1000
+	defaultTickTime = 100
+)
+
+type animationPlayMode byte
+
+const (
+	playModeForward animationPlayMode = iota
+	playModeBackword
+	playModeLeftRight
```
Comment (gravestench) [inline hswidget/dc6widget/state.go]: I think the name commonly given to this type of animation is `Yoyo`

Comment (gucio321) [reply hswidget/dc6widget/state.go]: ee0bc54 (renamed to pingPong, because of string of this constant)

**Review by essial — APPROVED** [review]

File: hswidget/dccwidget/state.go [hunk 4]
```diff
@@ -169,3 +258,52 @@ func (p *widget) makeImagePixel(val byte) color.RGBA {
...
+							if !state.repeat {
+								state.isPlaying = false
+							}
+						}
+					}
+				}
+			}
+		}
+	}
+}
```
Comment (gravestench) [inline hswidget/dccwidget/state.go]: I acknowledge your `nolint`, but I think 8 layers is too much...
```golang
// wrap integer to max: wrap(450, 360) == 90
func wrap(x, max int) int {
	wrapped := x % max

	if wrapped < 0 {
		return max + wrapped
	}

	return wrapped
}

func (p *widget) runPlayer(state *widgetState) {
	for range state.ticker.C {
		if !state.isPlaying {
			return
		}

		numFrames := int32(p.dcc.FramesPerDirection-1)
		isLastFrame := state.controls.frame == numFrames

		// update play direction
		switch state.playMode {
		case playModeForward:
			state.controls.leftRightDirection = true
		case playModeBackward:
			state.controls.leftRightDirection = false
		case playModeLeftRight:
			if isLastFrame {
				state.controls.leftRightDirection = !state.controls.leftRightDirection
			}
		}

		// now update the frame number
		isForward := state.controls.leftRightDirection == true

		if isForward {
			state.controls.frame++
		} else {
			state.controls.frame--
		}

		state.controls.frame = wrap(state.controls.frame, p.dcc.FramesPerDirection)

		// next, check for stopping/repeat
		isStoppingFrame := (state.controls.frame == 0) || (state.controls.frame == numFrames)

		if isStoppingFrame && !state.repeat {
			state.isPlaying = false
		}
	}
}
```

Comment (gucio321) [reply hswidget/dccwidget/state.go]: thanx!

**Pull request merged**